v2/proto/voucher.go

```diff
@@ -0,0 +1,139 @@
+package proto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type ReservationVoucher struct {
+	// Relay is the ID of the peer providing relay service
+	Relay peer.ID
+	// Peer is the ID of the peer receiving relay service through Relay
+	Peer peer.ID
+	// Expiration is the expiration time of the reservation
+	Expiration time.Time
+	// Signature is the signature of this voucher, as produced by the Relay peer
+	Signature []byte
```

Comment:
Can we just use our signed envelopes?

Reply:
Also, I'm _pretty_ sure we need the relays key for this to be useful. The envelope covers that, IIRC.

Reply:
Either that or we have to connect to the relay to get its key, which kind of makes the whole voucher thing pointless (at that point, we might as well just ask the relay "you relaying for peer X?").

Reply:
Yeah you are right, I will change it to use the envelope framework.

Reply:
done.